### PR TITLE
Feature/284 support for synchronous rest client using http client

### PR DESCRIPTION
**Pull request by suvmanda**
Implements issue #284

### CODE REVIEW

File: modules/http-client-rest/src/main/java/com/devonfw/module/httpclient/common/impl/rest/HttpRestClientSyncAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+
+import com.devonfw.module.service.common.api.client.sync.SyncHttpServiceClientFactory;
+import com.devonfw.module.service.common.impl.client.ServiceClientAutoConfiguration;
+
+/**
+ * {@link Configuration} for REST (JAX-RS) clients using Java HTTP client.
+ *
+ * @since 2020.08.001
```
Comment (ssarmokadam) [inline modules/http-client-rest/src/main/java/com/devonfw/module/httpclient/common/impl/rest/HttpRestClientSyncAutoConfiguration.java:13]: check from which version feature is available and modify javadoc accordingly

File: modules/http-client/src/main/java/com/devonfw/module/httpclient/common/impl/AbstractSyncServiceHttpClient.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  @SuppressWarnings("null")
+  @Override
+  protected <R> void doCall(ServiceClientInvocation<S> invocation, Consumer<R> resultHandler) {
+
+    long startTime = System.nanoTime();
+    HttpRequest request = createRequest(invocation);
+    HttpResponse<String> future = null;
+    try {
+      future = this.client.getHttpClient().send(request, BodyHandlers.ofString());
+    } catch (IOException e) {
```
Comment (ssarmokadam) [inline modules/http-client/src/main/java/com/devonfw/module/httpclient/common/impl/AbstractSyncServiceHttpClient.java:55]: Check for exception handling part in asynclient . do not add e.printStackTrace()

File: modules/service/src/main/java/com/devonfw/module/service/common/api/client/sync/SyncServiceClientFactory.java [hunk 3]
```diff
@@ -25,4 +25,14 @@
    */
   <S> S create(ServiceContext<S> context);
 
+  /**
```
Comment (ssarmokadam) [inline modules/service/src/main/java/com/devonfw/module/service/common/api/client/sync/SyncServiceClientFactory.java:28]: Remove commented code

**Review by ssarmokadam — CHANGES_REQUESTED** [review]
Check for javadocs on method and modify it for synchronous, check for versions in doc. Also check for error handling part

File: modules/service/src/main/java/com/devonfw/module/service/common/api/client/ServiceClientFactory.java [hunk 4]
```diff
@@ -79,4 +79,23 @@
...
+
+  <S> SyncServiceClient<S> createSync(Class<S> serviceInterface, Map<String, String> config);
+
+  /**
+   * @param <S> type of the {@link com.devonfw.module.service.common.api.client.context.ServiceContext#getApi() service
+   *        API}.
+   * @param serviceInterface the {@link Class} reflecting the interface that defines the API of your {@link Service}.
+   * @return an {@link SyncServiceClient} allowing to call service operations synchronously.
+   */
+  <S> SyncServiceClient<S> createSync(Class<S> serviceInterface);
```
Comment (hohwille) [inline modules/service/src/main/java/com/devonfw/module/service/common/api/client/ServiceClientFactory.java:99]: We already have this API. For historic reasons the methods are called `create`. When async support was added the `createAsync` methods had been added.
I do not see why we have to add `createSync` additionally that would IMHO do the same as the `create` methods. Do I miss something?

File: modules/service/src/main/java/com/devonfw/module/service/common/api/client/SyncServiceClient.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ * default like {@code 0} or {@code false}). This approach allows the {@link #call(Object, Consumer) call} method to be
+ * type-safe ensuring that the {@link Consumer} callback fits to the result type of the invoked operation.<br>
+ * Therefore ensure that you always call exactly one service operation on the {@link #get() service client} immediately
+ * followed by a {@link #call(Object, Consumer) call} method invocation.
+ *
+ * @param <S> type of the {@link #get() service client}.
+ * @since 2021.08.003
+ * @see ServiceClientFactory#createSync(Class)
+ */
+public interface SyncServiceClient<S> {
```
Comment (hohwille) [inline modules/service/src/main/java/com/devonfw/module/service/common/api/client/SyncServiceClient.java:31]: OK. It seems there is a missunderstanding here.
The synchronous access using HTTP should behave exactly like the API we already have existing.
Only the a new implementation option should be added that can be used instead without `cxf-client` dependencies...

IMHO for synchronous invocation this new API seems overcomplicated. I do not have to provide `Consumer` as callback when I can call the code directly.

**Review by hohwille — CHANGES_REQUESTED** [review]
@suvmanda Thanks for this PR. You have done a great job here. The code looks very well, has nice JavaDoc with `since` tags, code-style is fine. 👍 

However, you extended the API what seems to be a missunderstanding. Can you please check this and if you agree rework your PR?

File: modules/service/src/main/java/com/devonfw/module/service/common/api/client/SyncServiceClient.java (hunk 5 above)
Comment (hohwille) [reply modules/service/src/main/java/com/devonfw/module/service/common/api/client/SyncServiceClient.java:31]: See existing documentation:
https://github.com/devonfw/devon4j/blob/master/documentation/guide-service-client.asciidoc#simple-usage

```
 private void callSynchronous(MyArgs myArgs) {
    MyService myService = this.serviceClientFactory.create(MyService.class);
    // call of service over the wire, synchronously blocking until result is received or error occurred
    MyResult myResult = myService.myMethod(myArgs);
    handleResult(myResult);
  }
```

So when replacing dependency `devon4j-starter-cxf-client-rest` with `devon4j-starter-http-client-rest-sync` the same code pattern should work except that under the hood no more CXF but the HttpClient is used.

File: modules/http-client-rest/src/main/java/com/devonfw/module/httpclient/common/impl/rest/SyncServiceClientStubImpl.java [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+  private Consumer<Throwable> errorHandler;
+
+  /** The most recent invocation. */
+  private ServiceClientInvocation<S> invocation;
+
+  private ServiceClientStub<S> stub;
+
+  /** The owning {@link SyncServiceClientFactoryHttp factory} which created this client. */
+  // protected final F factory = null;
+
```
Comment (hohwille) [inline modules/http-client-rest/src/main/java/com/devonfw/module/httpclient/common/impl/rest/SyncServiceClientStubImpl.java:63]: remove outcommented code.

File: modules/http-client-rest/src/main/java/com/devonfw/module/httpclient/common/impl/rest/SyncServiceClientStubImpl.java [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+    String contentType = response.headers().firstValue("Content-Type").orElse("application/json");
+    String data = "";
+    Object body = response.body();
+    if (body instanceof String) {
+      data = (String) body;
+    } else {
+      handleUnsupportedBody(body);
+    }
+    // return this.factory.getErrorUnmarshaller().unmarshall(data, contentType, statusCode, service);
+    return null;
```
Comment (hohwille) [inline modules/http-client-rest/src/main/java/com/devonfw/module/httpclient/common/impl/rest/SyncServiceClientStubImpl.java:139]: Please complete this with proper code.

File: modules/http-client-rest/src/main/java/com/devonfw/module/httpclient/common/impl/rest/SyncServiceClientStubImpl.java [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+  /** {@link ServiceHttpClient} to use. */
+  protected final ServiceHttpClient client;
+
+  private final ObjectMapper objectMapper;
+
+  private final RestServiceMetadata<S> serviceMetadata;
+
+  /** The {@link #setErrorHandler(Consumer)} */
+  private Consumer<Throwable> errorHandler;
```
Comment (hohwille) [inline modules/http-client-rest/src/main/java/com/devonfw/module/httpclient/common/impl/rest/SyncServiceClientStubImpl.java:54]: should also be removed as this only  makes sense for async usage. There is no API to set this in sync usage and the sync API is designed to throw errors immediately when received.

File: modules/http-client/src/main/java/com/devonfw/module/httpclient/common/impl/AbstractSyncServiceHttpClient.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import com.devonfw.module.service.common.base.client.AbstractSyncServiceClient;
+
+/**
+ * Abstract base implementation of {@link SyncServiceClient} using Java HTTP client.
+ *
+ * @param <S> type of the {@link #get() service client}.
+ * @param <F> type of the owning {@link SyncServiceClientFactoryHttp factory}.
+ * @since 2021.08.003
+ */
+
```
Comment (hohwille) [inline modules/http-client/src/main/java/com/devonfw/module/httpclient/common/impl/AbstractSyncServiceHttpClient.java:18]: can you remove all these pointless empty lines between JavaDoc and Signature, please?

**Review by hohwille — CHANGES_REQUESTED** [review]
@suvmanda thanks for your updates. We are getting close. Please have a look at my review comments.
There has been some C&P from the existing code. I can resolve that myself after merging.

**Comment by hohwille** [comment]
@suvmanda thanks again for your work and this PR. I forked your work and created PR #410 that replaces this PR.
Therefore I am closing this PR.

**Pull request closed without merging**